File: Cisco/GitHubList/Python/9ValidParentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        closed_opened = {
            "]": "[",
            "}": "{",
            ")": "(",
        }
        lifo = []
        for br in s:
            if br in ("[", "(", "{"):
                lifo.append(br)
            else:
                if len(lifo) == 0 or closed_opened[br] != lifo.pop():
                    return False

        return True if len(lifo) == 0 else False

File: Cisco/GitHubList/C++/9ValidParentheses.cpp
#include <string>
#include <stack>

using namespace std;

class Solution
{
    constexpr char getBackward(char forward)
    {
        switch (forward)
        {
        case ')':
            return '(';
        case '}':
            return '{';
        case ']':
            return '[';
        default:
            return '\000';
        }
    }

public:
    bool isValid(string s)
    {
        stack<char> parentheses;

        for (const auto &ch : s)
            if (ch == ')' || ch == '}' || ch == ']')
                if (parentheses.size() == 0 || getBackward(ch) != parentheses.top())
                    return false;
                else
                    parentheses.pop();
            else
                parentheses.push(ch);

        return parentheses.size() == 0;
    }
};